index.js

```diff
@@ -2,188 +2,202 @@
-        }
+    if (dependencies.length) {
+      cyclicDependencyCheck[name] = true
+      dependencies.forEach((name) => allPlugins[name] && loadPlugin(allPlugins[name]))
+    }
 
-        registerPlugin(name, plugin, options)
-      }
+    registerPlugin(name, plugin, options)
+  }
+
+  const pluginKeys = Object.keys(allPlugins)
+  for (let i = 0; i < pluginKeys.length; i++) {
+    cyclicDependencyCheck = {}
+
+    try {
+      loadPlugin(allPlugins[pluginKeys[i]])
+    } catch (err) {
+      throw enrichError(err)
+    }
+  }
+}
+
+// do not create a new context, do not encapsulate
+// same as fastify-plugin
+module.exports[Symbol.for('skip-override')] = true
+
+async function loadPluginFiles (opts) {
+  const packagePattern = /^package\.json$/im
+  const indexPattern = opts.includeTypeScript
+    ? /^index\.(ts|js|mjs)$/im
+    : /^index\.(js|mjs)$/im
+  const scriptPattern = opts.includeTypeScript
+    ? /\.(ts|js|mjs)$/im
+    : /\.(js|mjs)$/im
 
-      const pluginKeys = Object.keys(allPlugins)
-      for (let i = 0; i < pluginKeys.length; i++) {
-        cyclicDependencyCheck = {}
+  const list = await readDir(opts.dir)
 
-        try {
-          loadPlugin(allPlugins[pluginKeys[i]])
-        } catch (err) {
-          next(enrichError(err))
-          return
+  const results = []
+  for (const file of list) {
+    if (opts.ignorePattern && file.match(opts.ignorePattern)) {
+      results.push({ skip: true })
+      continue
+    }
+    const toLoad = path.join(opts.dir, file)
+    const stat = await getStat(toLoad)
+
+    if (stat.isDirectory()) {
+      const files = await readDir(toLoad)
+
+      const fileList = files.join('\n')
+      // if the directory does not contain a package.json or an index,
+      // load each script file as an independend plugin
+      if (
+        !packagePattern.test(fileList) &&
+        !indexPattern.test(fileList) &&
+        scriptPattern.test(fileList)
+      ) {
+        const plugins = []
+        for (let index = 0; index < files.length; index++) {
+          const file = files[index]
+
+          plugins.push({
+            skip: !scriptPattern.test(file),
+            opts: {
+              prefix: toLoad.split(path.sep).pop()
+            },
+            file: path.join(toLoad, file)
+          })
         }
+        results.push(plugins)
+      } else {
+        results.push({
+          // skip directories without script files inside
+          skip: files.every(name => !scriptPattern.test(name)),
+          file: toLoad
+        })
+      }
+    } else {
+      results.push({
+        // only accept script files
+        skip: !(stat.isFile() && scriptPattern.test(file)),
+        file: toLoad
+      })
+    }
+  }
+
+  return results
+}
+
+function getStat (toLoad) {
+  return new Promise((resolve, reject) => {
+    fs.stat(toLoad, (err, stat) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(stat)
       }
+    })
+  })
+}
```

Comment:
Can you please use `util.promisify` instead of manually wrapping them?

Reply:
I used `es6-promisify` for now since node 6 has no `util.promisify`: 
https://nodejs.org/docs/latest-v6.x/api/util.html